OpenRA.Mods.Common/Traits/TargetableUnit.cs

```diff
@@ -52,5 +63,26 @@ public virtual IEnumerable<WPos> TargetablePositions(Actor self)
 		}
 
 		public bool RequiresForceFire { get { return info.RequiresForceFire; } }
+
+		public IEnumerable<string> UpgradeTypes	{ get { return info.UpgradeTypes; } }
+
+		public bool AcceptsUpgradeLevel(Actor self, string type, int level)
+		{
+			return level <= 1;
+		}
+
+		public void UpgradeLevelChanged(Actor self, string type, int oldLevel, int newLevel)
+		{
+			if (newLevel > 0)
+			{
+				if (info.UpgradedTargetTypes.Contains("!"))
+					targetTypes = info.UpgradedTargetTypes.Where(utt => !utt.StartsWith("!")).ToArray();
+				else
+					targetTypes = targetTypes.Except(info.UpgradedTargetTypes.Where(utt => utt.StartsWith("!")).Select(utt => utt.Substring(1)))
```

Comment:
No, that's correct. The first `if` checks whether a *single* ! exists in `UpgradedTargetTypes`. In that case, `TargetTypes` will be *replaced* with all entries in `UpgradedTargetTypes` that don't start with an !.

The `else` case instead *modifies* the existing `TargetTypes` and removes all entries that start with an !, and adds all others.